config/database.go

```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package config
+
+import (
+	"bytes"
+	"database/sql"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+
+	// Load the MySQL driver
+	_ "github.com/go-sql-driver/mysql"
+	// Load the Postgres driver
+	_ "github.com/lib/pq"
+)
+
+// DatabaseStore is a config store backed by a database.
+type DatabaseStore struct {
+	emitter
+
+	config               *model.Config
+	environmentOverrides map[string]interface{}
+	configLock           sync.RWMutex
+	originalDsn          string
+	driverName           string
+	dataSourceName       string
+	db                   *sqlx.DB
+}
+
+// NewDatabaseStore creates a new instance of a config store backed by the given database.
+func NewDatabaseStore(dsn string) (ds *DatabaseStore, err error) {
+	driverName, dataSourceName, err := parseDSN(dsn)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid DSN")
+	}
+
+	// TODO: Retry logic?
```

Comment:
This happens at the very beginning of the start of the application, if the application runs, this works, if I can't connect to the database, I can fail gracefully or I can retry some times, but the retry can be done here, in a upper level, or maybe as part of the systemd script used to run mattermost in  your machine.

I would suggest to keep it as simple as is now, if can connect great, if not fail.